### ISSUE TITLE
Permission for Viewing Revisions

### ISSUE THREAD

**Issue by 2devs1code** [BookStackApp/BookStack#4526]
### Describe the feature you'd like

I would like to have the possibility to remove the ability for anyone in the world to view the revisions of a page. Put another way, I would like for revision viewing to be a specific permission, rather than tied to the blanket ability to view a page at all.

### Describe the benefits this would bring to existing BookStack users

Sometimes mistakes are made to previous revisions or directions are given that may prove dangerous necessitating the updating of the pages to correct the mistake or remove the dangerous information.

While I can remove revisions now, I must do it by hand or scripting with PHP artisan and I don't necessarily want to do a blanket removal of revisions. While I want my admins and documentation people to be able to see and use revisions, I can't see a good reason for the general public to see the old revisions.

### Can the goal of this request already be achieved via other means?

Unknown. Hiding the links would be sufficient too, though permissions are preferred. The old -and now unavailable- behavior of requiring edit to view revisions was also sufficient.

### Have you searched for an existing open/closed issue?

- [X] I have searched for existing issues and none cover my fundemental request

### How long have you been using BookStack?

6 months to 1 year

**Comment by maxist0** [comment]
One simple solution for you would be to change the code yourself. You just need to edit app/Entities/Controllers/PageRevisionController.php
You have to call:  `$this->checkPermission('settings-manage'); ` in every method (index, show, changes). By doing this only admin-user can see the revision viewing.

**Comment by Man-in-Black** [comment]
Just to mention: I would do this as part of the visual theme system and not by editing the core files, because these files are getting replaced on an update.

**Comment by maxist0** [comment]
Not exactly. The visual theme perhaps will get replaced too on an update. You also don't want to just remove the link to the revision page itself, because it is still accessible over URL. Best way to handle an update would be a Git rebase or so

**Comment by ssddanbrown** [comment]
The visual theme system API is deemed stable, so would be documented/announced if changed (and it's something I try to ensure remains stable), but the contents of views are not. You could add permission checks into views to hide content on permission check, but there'd be a lot to add.

Using the [logical theme system](https://github.com/BookStackApp/BookStack/blob/development/dev/docs/logical-theme-system.md) you could do something like this to add prevention without altering app files:

```php
<?php

use BookStack\Exceptions\NotifyException;
use BookStack\Facades\Theme;
use BookStack\Theming\ThemeEvents;
use Illuminate\Http\Request;

Theme::listen(ThemeEvents::WEB_MIDDLEWARE_BEFORE, function (Request $request) {
    if (str_contains($request->url(), '/revisions') && !userCan('settings-manage')) {
        throw new NotifyException("Access to revisions not allowed", '/', 403);
    }
});
```

_(Hacks like the above are unsupported and unofficial functionality)_

**Comment by ssddanbrown** [comment]
I'm re-opening since I've now had a good few requests for this so would be something to potentially include as a permission in control.

Additional request and context in  #4544.

**Comment by FAFDK** [comment]
Any updates on this? Would like to hide revisions in public. Thank you for your work.

**Comment by byPavel** [comment]
I would like to get this feature too

**Comment by seclution** [comment]
Would be very nice to see that feature :)

**Comment by aberkvam** [comment]
+1

For the time being, we're setting REVISIONS_LIMIT=0 but it would be nice to keep the revisions feature.

**Comment by lxwulf** [comment]
+1

> @aberkvam thank you for mentioning this option from/for the `.env` file. 🤝

**Comment by mmullikinLG** [comment]
Please add this feature.

**Comment by BStenfors9119** [comment]
+1

**Comment by ssddanbrown** [comment]
The changes in #6108 introduce a new role-level "Revisions - View-all" permission which can (globally) control access to revision-related views and actions.

These changes will be part of the next feature release.
By default, on upgrade, any roles with any role-level page-view permission will be given this permission to provide a forward compatible upgrade path.
